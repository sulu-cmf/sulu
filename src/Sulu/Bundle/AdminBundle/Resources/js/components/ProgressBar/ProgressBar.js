// @flow
import React from 'react';
import classNames from 'classnames';
import styles from './progressBar.scss';

type Props = {
    max: number,
    style: 'progress' | 'success' | 'error' | 'warning',
    value: number,
}

class ProgressBar extends React.PureComponent<Props> {
    static defaultProps = {
        style: 'progress',
    };

    render() {
        const {value, max, style} = this.props;

        const className = classNames(styles.progressBar, styles[style]);

        return (
            <progress className={className} max={max} value={value}>
                {(value / max) * 100}%
            </progress>
        );
    }
}

export default ProgressBar;
