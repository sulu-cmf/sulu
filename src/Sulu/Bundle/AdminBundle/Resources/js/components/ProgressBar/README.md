A simple progress bar component used to indicate the progress of multiple tasks or processes.

```javascript
<ProgressBar
    value={50}
    max={100}
/>
```

The style can be customized.
Beside the default `progress` style, it supports a `success` style,

```javascript
<ProgressBar
    value={100}
    max={100}
    type="success"
/>
```

an `error` style

```javascript
<ProgressBar
    value={25}
    max={100}
    type="error"
/>
```

and a `warning` style.

```javascript
<ProgressBar
    value={75}
    max={100}
    type="warning"
/>
```
