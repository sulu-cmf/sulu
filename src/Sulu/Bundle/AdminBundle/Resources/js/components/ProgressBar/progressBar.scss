@import '../../containers/Application/colors.scss';

$width: 250px;
$height: 20px;
$borderRadius: 3px;
$backgroundColor: $white;
$borderColor: $silver;
$progressColor: $shakespeare;
$progressColorError: $roman;
$progressColorWarning: $gold;
$progressColorSuccess: $mantis;

.progressBar {
    appearance: none;
    width: $width;
    height: $height;
    overflow: hidden;
    border: 1px solid $borderColor;
    border-radius: $borderRadius;
    background-color: $backgroundColor;

    --progress-color: $progressColor;

    &.error {
        --progress-color: $progressColorError;
    }

    &.warning {
        --progress-color: $progressColorWarning;
    }

    &.success {
        --progress-color: $progressColorSuccess;
    }

    &::-webkit-progress-bar {
        overflow: hidden;
        background-color: $backgroundColor;
    }

    &::-webkit-progress-value {
        background-image:
            -webkit-linear-gradient(
                -45deg,
                transparent 33%,
                rgba(0, 0, 0, .1) 33%,
                rgba(0, 0, 0, .1) 66%,
                transparent 66%
            ),
            -webkit-linear-gradient(
                top,
                rgba(255, 255, 255, .25),
                rgba(0, 0, 0, .25)
            ),
            -webkit-linear-gradient(
                left,
                var(--progress-color),
                var(--progress-color)
            );
        border-radius: $borderRadius;
        background-size: 35px 20px, 100% 100%, 100% 100%;
        transform: scale(1.05, 1.05);
    }

    &::-moz-progress-bar {
        background-image:
            -moz-linear-gradient(
                135deg,
                transparent 33%,
                rgba(0, 0, 0, .1) 33%,
                rgba(0, 0, 0, .1) 66%,
                transparent 66%
            ),
            -moz-linear-gradient(
                top,
                rgba(255, 255, 255, .25),
                rgba(0, 0, 0, .25)
            ),
            -moz-linear-gradient(
                left,
                var(--progress-color),
                var(--progress-color)
            );
        border-radius: $borderRadius;
        background-size: 35px 20px, 100% 100%, 100% 100%;
        transform: scale(1.05, 1.05);
    }
}
