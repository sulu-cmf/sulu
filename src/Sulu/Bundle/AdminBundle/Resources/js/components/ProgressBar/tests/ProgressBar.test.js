// @flow
import {render, mount} from 'enzyme';
import React from 'react';
import ProgressBar from '../ProgressBar';

test('The component should render', () => {
    const view = render(
        <ProgressBar
            max={10}
            value={7}
        />
    );

    expect(view).toMatchSnapshot();
});

test('The component should render with progress skin as default', () => {
    const view = mount(
        <ProgressBar
            max={10}
            value={5}
        />
    );

    expect(view.find('.progressBar').prop('max')).toBe(10);
    expect(view.find('.progressBar').prop('value')).toBe(5);
    expect(view.find('.progressBar').hasClass('progress'));
    expect(view.find('.progressBar').text()).toBe('50%');
});

test('The component should render with success skin', () => {
    const view = mount(
        <ProgressBar
            max={10}
            skin="success"
            value={10}
        />
    );

    expect(view.find('.progressBar').prop('max')).toBe(10);
    expect(view.find('.progressBar').prop('value')).toBe(10);
    expect(view.find('.progressBar').hasClass('success'));
    expect(view.find('.progressBar').text()).toBe('100%');
});

test('The component should render with warning skin', () => {
    const view = mount(
        <ProgressBar
            max={10}
            skin="warning"
            value={0}
        />
    );

    expect(view.find('.progressBar').prop('max')).toBe(10);
    expect(view.find('.progressBar').prop('value')).toBe(0);
    expect(view.find('.progressBar').hasClass('warning'));
    expect(view.find('.progressBar').text()).toBe('0%');
});

test('The component should render with error skin', () => {
    const view = mount(
        <ProgressBar
            max={10}
            skin="error"
            value={3}
        />
    );

    expect(view.find('.progressBar').prop('max')).toBe(10);
    expect(view.find('.progressBar').prop('value')).toBe(3);
    expect(view.find('.progressBar').hasClass('error'));
    expect(view.find('.progressBar').text()).toBe('30%');
});

test('The component should render with max 0', () => {
    const view = mount(
        <ProgressBar
            max={0}
            skin="error"
            value={0}
        />
    );

    expect(view.find('.progressBar').prop('max')).toBe(1);
    expect(view.find('.progressBar').prop('value')).toBe(0);
    expect(view.find('.progressBar').hasClass('error'));
    expect(view.find('.progressBar').text()).toBe('0%');
});
