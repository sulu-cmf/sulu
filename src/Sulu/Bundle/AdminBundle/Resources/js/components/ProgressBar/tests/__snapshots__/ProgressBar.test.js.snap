// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The component should render 1`] = `
<progress
  class="progressBar progress"
  max="10"
  value="7"
>
  70%
</progress>
`;
